File: OsmAndWear/src/net/osmand/plus/configmap/tracks/TrackFolderLoaderTask.java
package net.osmand.plus.configmap.tracks;

import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import androidx.annotation.WorkerThread;

import net.osmand.PlatformUtil;
import net.osmand.plus.OsmandApplication;
import net.osmand.shared.gpx.GpxDataItem;
import net.osmand.shared.gpx.GpxDbHelper;
import net.osmand.shared.gpx.GpxDbHelper.GpxDataItemCallback;
import net.osmand.shared.gpx.GpxHelper;
import net.osmand.shared.gpx.SmartFolderHelper;
import net.osmand.shared.gpx.TrackItem;
import net.osmand.shared.gpx.data.TrackFolder;
import net.osmand.shared.io.KFile;
import net.osmand.util.Algorithms;

import org.apache.commons.logging.Log;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

@Deprecated
public class TrackFolderLoaderTask extends AsyncTask<Void, TrackItem, Void> {

	private static final Log LOG = PlatformUtil.getLog(TrackFolderLoaderTask.class);
	private static final int LOG_BATCH_SIZE = 100;
	private static final int PROGRESS_BATCH_SIZE = 70;

	private final GpxDbHelper gpxDbHelper;
	private final SmartFolderHelper smartFolderHelper;

	private final TrackFolder folder;
	private final LoadTracksListener listener;
	private long loadingTime = 0;
	private int tracksCounter = 0;
	private GpxDataItemCallback callback;

	public TrackFolderLoaderTask(@NonNull OsmandApplication app, @NonNull TrackFolder folder, @NonNull LoadTracksListener listener) {
		this.folder = folder;
		this.listener = listener;
		this.gpxDbHelper = app.getGpxDbHelper();
		this.smartFolderHelper = app.getSmartFolderHelper();
	}

	@Override
	protected void onPreExecute() {
		if (listener != null) {
			listener.loadTracksStarted();
		}
	}

	@Override
	protected void onProgressUpdate(TrackItem... items) {
		if (listener != null) {
			listener.loadTracksProgress(items);
		}
	}

	@Override
	protected Void doInBackground(Void... voids) {
		long start = System.currentTimeMillis();
		LOG.info("Start loading tracks in " + folder.getDirName());

		folder.clearData();
		loadingTime = System.currentTimeMillis();

		List<TrackItem> progress = new ArrayList<>();
		loadGPXFolder(folder, progress);

		if (!progress.isEmpty()) {
			publishProgress(progress.toArray(new TrackItem[0]));
		}
		if (listener != null) {
			listener.tracksLoaded(folder);
		}
		LOG.info("Finished loading tracks. Took " + (System.currentTimeMillis() - start) + "ms");
		return null;
	}

	private void loadGPXFolder(@NonNull TrackFolder rootFolder, @NonNull List<TrackItem> progress) {
		Deque<TrackFolder> folders = new ArrayDeque<>();
		folders.push(rootFolder);

		while (!folders.isEmpty()) {
			TrackFolder folder = folders.pop();
			KFile dir = folder.getDirFile();
			List<KFile> files = dir.listFiles();
			if (Algorithms.isEmpty(files)) {
				continue;
			}
			List<TrackItem> trackItems = new ArrayList<>();
			List<TrackFolder> subFolders = new ArrayList<>();

			for (KFile file : files) {
				if (isCancelled()) {
					return;
				}
				if (file.isDirectory()) {
					TrackFolder subfolder = new TrackFolder(file, folder);
					subFolders.add(subfolder);
					folders.push(subfolder); // Add subfolder to the queue for processing
				} else if (GpxHelper.INSTANCE.isGpxFile(file)) {
					TrackItem item = new TrackItem(file);
					item.setDataItem(getDataItem(item, file));
					trackItems.add(item);

					progress.add(item);
					if (progress.size() > PROGRESS_BATCH_SIZE) {
						publishProgress(progress.toArray(new TrackItem[0]));
						progress.clear();
					}
					tracksCounter++;
					if (tracksCounter % LOG_BATCH_SIZE == 0) {
						long endTime = System.currentTimeMillis();
						LOG.info("Loading " + LOG_BATCH_SIZE + " tracks. Took " + (endTime - loadingTime) + "ms");
						loadingTime = endTime;
					}
				}
			}
			folder.setTrackItems(trackItems);
			folder.setSubFolders(subFolders);
			smartFolderHelper.addTrackItemsToSmartFolder(trackItems);
		}
		for (TrackFolder folder : rootFolder.getFlattenedSubFolders()) {
			folder.resetCachedData();
		}
		rootFolder.resetCachedData();
	}

	@Nullable
	private GpxDataItem getDataItem(@NonNull TrackItem trackItem, @NonNull KFile file) {
		if (callback == null) {
			callback = new GpxDataItemCallback() {
				@Override
				public boolean isCancelled() {
					return TrackFolderLoaderTask.this.isCancelled();
				}

				@Override
				public void onGpxDataItemReady(@NonNull GpxDataItem item) {
					trackItem.setDataItem(item);
				}
			};
		}
		return gpxDbHelper.getItem(file, callback);
	}

	@Override
	protected void onPostExecute(Void unused) {
		if (listener != null) {
			listener.loadTracksFinished(folder);
		}
		smartFolderHelper.notifyUpdateListeners();
	}

	public interface LoadTracksListener {

		@UiThread
		default void loadTracksStarted() {
		}

		@UiThread
		default void loadTracksProgress(@NonNull TrackItem[] items) {
		}

		@WorkerThread
		default void tracksLoaded(@NonNull TrackFolder folder) {
		}

		@UiThread
		void loadTracksFinished(@NonNull TrackFolder folder);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/gpx/TrackFolderLoaderTask.kt
package net.osmand.shared.gpx

import co.touchlab.stately.concurrency.Synchronizable
import co.touchlab.stately.concurrency.synchronize
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.osmand.shared.KAsyncTask
import net.osmand.shared.extensions.currentTimeMillis
import net.osmand.shared.gpx.GpxDbHelper.GpxDataItemCallbackEx
import net.osmand.shared.gpx.data.TrackFolder
import net.osmand.shared.io.KFile
import net.osmand.shared.util.LoggerFactory

class TrackFolderLoaderTask(
	private val folder: TrackFolder,
	private val listener: LoadTracksListener,
	private val forceLoad: Boolean = false
) : KAsyncTask<Unit, TrackItem, TrackFolder>(true) {

	companion object {
		private val log = LoggerFactory.getLogger("TrackFolderLoaderTask")

		private const val LOG_BATCH_SIZE = 100
		private const val PROGRESS_BATCH_SIZE = 70

		private var cachedRootFolder: TrackFolder? = null
		private var cachedDbItemsVersion = -1
	}

	private var loadingTime = 0L
	private var tracksCounter = 0
	private var taskSync = Synchronizable()
	private var progressSync = Synchronizable()

	private fun shouldLoadFolder(cachedRootFolder: TrackFolder?) =
		forceLoad || cachedRootFolder == null || !folder.isRootFolder || cachedDbItemsVersion != GpxDbHelper.getItemsVersion()

	override fun onPreExecute() {
		if (shouldLoadFolder(cachedRootFolder)) listener.loadTracksStarted()
	}

	override fun onProgressUpdate(vararg values: TrackItem) {
		if (shouldLoadFolder(cachedRootFolder)) listener.loadTracksProgress(*values)
	}

	override suspend fun doInBackground(vararg params: Unit): TrackFolder {
		val cachedRootFolder = cachedRootFolder
		if (!shouldLoadFolder(cachedRootFolder)) return cachedRootFolder!!

		val start = currentTimeMillis()
		log.info("Start loading tracks in ${folder.getDirName()}")

		folder.clearData()
		loadingTime = currentTimeMillis()

		val progress = mutableListOf<TrackItem>()
		loadGPXFolder(folder, progress)
		if (folder.isRootFolder) {
			Companion.cachedRootFolder = TrackFolder(folder)
			cachedDbItemsVersion = GpxDbHelper.getItemsVersion()
		}

		listener.tracksLoaded(folder)
		log.info("Finished loading tracks. Took ${currentTimeMillis() - start}ms")

		return folder
	}

	override fun onPostExecute(result: TrackFolder) {
		val cachedRootFolder = cachedRootFolder
		if (result === cachedRootFolder) folder.update(cachedRootFolder)
		resetCachedData(folder)
		listener.loadTracksFinished(folder)
		SmartFolderHelper.notifyUpdateListeners()
	}

	private suspend fun loadGPXFolder(
		rootFolder: TrackFolder,
		progress: MutableList<TrackItem>
	) {
		val trackItems = mutableListOf<TrackItem>()
		withContext(Dispatchers.IO) {
			scanFolder(rootFolder, progress, trackItems)
		}
		SmartFolderHelper.addTrackItemsToSmartFolder(trackItems)

		resetCachedData(rootFolder)
	}

	private fun resetCachedData(rootFolder: TrackFolder) {
		for (subFolder in rootFolder.getFlattenedSubFolders()) subFolder.resetCachedData()
		rootFolder.resetCachedData()
	}

	private suspend fun scanFolder(
		folder: TrackFolder,
		progress: MutableList<TrackItem>,
		trackItems: MutableList<TrackItem>
	): Unit = coroutineScope {
		val dir = folder.getDirFile()
		val files = dir.listFiles()
		if (files.isNullOrEmpty()) return@coroutineScope

		val fileBatches: List<List<KFile>> = files.chunked(PROGRESS_BATCH_SIZE)
		fileBatches.forEach { batch ->
			launch {
				log.info("Loading track file batch = ${batch.size}")
				val batchTrackItems = mutableListOf<TrackItem>()
				for (file in batch) {
					if (isCancelled()) return@launch

					if (file.isDirectory()) {
						val subfolder = TrackFolder(file, folder)
						launch {
							log.info("Loading track subfolder = $subfolder")
							scanFolder(subfolder, progress, trackItems)
							taskSync.synchronize { folder.addSubFolder(subfolder) }
						}
					} else if (GpxHelper.isGpxFile(file)) {
						val item = TrackItem(file)
						item.dataItem = getDataItem(item, file)
						taskSync.synchronize { trackItems.add(item) }
						batchTrackItems.add(item)

						progressSync.synchronize {
							progress.add(item)
							if (progress.size > PROGRESS_BATCH_SIZE) {
								publishProgress(*progress.toTypedArray())
								progress.clear()
							}

							tracksCounter++
							if (tracksCounter % LOG_BATCH_SIZE == 0) {
								val endTime = currentTimeMillis()
								log.info("Loading $LOG_BATCH_SIZE tracks. Took ${endTime - loadingTime}ms")
								loadingTime = endTime
							}
						}
					}
				}
				taskSync.synchronize { folder.addTrackItems(batchTrackItems) }
			}
		}
		if (isCancelled()) return@coroutineScope
	}

	private fun getDataItem(trackItem: TrackItem, file: KFile): GpxDataItem? {
		val callback = object : GpxDataItemCallbackEx {
			override fun isCancelled(): Boolean = this@TrackFolderLoaderTask.isCancelled()

			override fun onGpxDataItemReady(item: GpxDataItem) {}

			override fun onGpxDataItemReady(item: GpxDataItem, lastItem: Boolean) {
				if (file == item.file) {
					trackItem.dataItem = item
				}
				if (lastItem && !isRunning()) {
					resetCachedData(folder)
					listener.deferredLoadTracksFinished(folder)
				}
			}
		}
		return GpxDbHelper.getItem(file, callback)
	}

	fun interface LoadTracksListener {
		fun loadTracksStarted() {}
		fun loadTracksProgress(vararg items: TrackItem) {}
		fun tracksLoaded(folder: TrackFolder) {}
		fun loadTracksFinished(folder: TrackFolder)
		fun deferredLoadTracksFinished(folder: TrackFolder) {}
	}
}